src/natcap/invest/utils.py

```diff
@@ -597,119 +597,127 @@ def expand_path(path, base_path):
+            usecols=lambda col: any(
+                re.fullmatch(pattern, col.strip().lower()) for pattern in patterns
+            ),
+            **{
+                'sep': None,
+                'engine': 'python',
+                'encoding': 'utf-8-sig',
+                **kwargs
+            })
     except UnicodeDecodeError as error:
         LOGGER.error(
             f'The file {path} must be encoded as UTF-8 or ASCII')
         raise error
 
-    # strip whitespace from column names
+    # strip whitespace from column names and convert to lowercase
     # this won't work on integer types, which happens if you set header=None
     # however, there's little reason to use this function if there's no header
-    dataframe.columns = dataframe.columns.str.strip()
-
-    # convert column names to lowercase
-    if convert_cols_to_lower:
-        dataframe.columns = dataframe.columns.str.lower()
-        # if 'to_lower`, case handling is done before trying to access the data.
-        # the columns are stripped of leading/trailing whitespace in
-        # ``read_csv_to_dataframe``, and also lowercased if ``to_lower`` so we only
-        # need to convert the rest of the table.
-        if index_col and isinstance(index_col, str):
-            index_col = index_col.lower()
-        # lowercase column names
-        if usecols:
-            usecols = [col.lower() for col in usecols]
-
-    # Subset dataframe by columns if desired
-    if usecols:
-        dataframe = dataframe[usecols]
-
-    # Set 'index_col' as the index of the dataframe
-    if index_col:
+    df.columns = df.columns.str.strip().str.lower()
+
+    # drop any empty rows
+    df = df.dropna(how="all")
+
+    available_cols = set(df.columns)
+
+    for col_spec, pattern in zip(spec['columns'].values(), patterns):
+        matching_cols = [c for c in available_cols if re.match(pattern, c)]
+        available_cols -= set(matching_cols)
+        for col in matching_cols:
+            try:
+                if col_spec['type'] in ['csv', 'directory', 'file', 'raster', 'vector', {'vector', 'raster'}]:
+                    df[col] = df[col].apply(
+                        lambda p: p if pandas.isna(p) else expand_path(str(p).strip(), path))
+                    df[col] = df[col].astype(pandas.StringDtype())
+                elif col_spec['type'] in {'freestyle_string', 'option_string'}:
+                    df[col] = df[col].apply(
+                        lambda s: s if pandas.isna(s) else str(s).strip().lower())
+                    df[col] = df[col].astype(pandas.StringDtype())
+                elif col_spec['type'] in {'number', 'percent', 'ratio'}:
+                    df[col] = df[col].astype(float)
+                elif col_spec['type'] == 'integer':
+                    df[col] = df[col].astype(pandas.Int64Dtype())
+                elif col_spec['type'] == 'boolean':
+                    df[col] = df[col].astype('boolean')
+            except ValueError as err:
```

Comment:
Would it be worth adding an `else` clause here?  Or are we assuming that any other type will be caught by the MODEL_SPEC tests?

Reply:
Yeah, since this is only being used internally by invest models, I think we can be sure that we'll only encounter valid types